### ISSUE TITLE
delays in receiving SCP messages from connected peers

### ISSUE THREAD

**Issue by MonsieurNicolas** [stellar/stellar-core#2441]
I don't know if this is a different incarnation of #2368 

On watcher node that I have:
```
2020-02-20T22:01:58.365 GABA5 [Ledger INFO] Closed ledger: [seq=28319360, hash=4a23ad]
2020-02-20T22:02:09.164 GABA5 [Herder INFO] Quorum information for 28319359 : {"agree":3,"delayed":0,"disagree":0,"fail_at":2,"hash":"75b44e","ledger":28319359,"missing":0,"phase":"EXTERNALIZE"}

2020-02-20T22:02:09.168 GABA5 [Ledger INFO] Got consensus: [seq=28319361, prev=4a23ad, txs=117, ops=234, sv: [ txH: 232f01, ct: 1582236123, upgrades: [ ] ]]
2020-02-20T22:02:09.175 GABA5 [Tx INFO] applying ledger 28319361 (txs:117, ops:234)
2020-02-20T22:02:09.236 GABA5 [Ledger INFO] Closed ledger: [seq=28319361, hash=f8ac19]
```

Some messages that the watcher node saw:
```
            [
               1582236123,
               "{ENV@GAAV2 |  i: 28319361 | NOMINATE | D: 341a0a | X: {'[ SIGNED@GAAV2 txH: 232f01, ct: 1582236123, upgrades: [ ] ]'} | Y: {} }",
               false
            ],
...
[
               1582236125,
               "{ENV@GDKWE |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ],
            [
               1582236125,
               "{ENV@GA35T |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ],
            [
               1582236125,
               "{ENV@GCFON |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ],
            [
               1582236125,
               "{ENV@GD6SZ |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ],
            [
               1582236125,
               "{ENV@GADLA |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ],
            [
               1582236125,
               "{ENV@GAZ43 |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ],
            [
               1582236125,
               "{ENV@GDXQB |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ],
[
               1582236129,
               "{ENV@sdf_live_1 |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ],
            [
               1582236129,
               "{ENV@sdf_live_2 |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ],
            [
               1582236129,
               "{ENV@sdf_live_3 |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               false
            ]
```
Round started at 1582236123, most nodes sent an EXTERNALIZE message at 1582236125 but the SDF nodes sent their EXTERNALIZE message at 1582236129.

Looking at the SDF validators, I saw this:
```
2020-02-20T22:01:58.314 GCGB2 [Ledger INFO] Got consensus: [seq=28319360, prev=ed90cf, txs=110, ops=219, sv: [ txH: 856ebb, ct: 1582236117, upgrades: [ ] ]]
2020-02-20T22:01:58.333 GCGB2 [Tx INFO] applying ledger 28319360 (txs:110, ops:219)
2020-02-20T22:01:58.461 GCGB2 [History INFO] Publishing 1 queued checkpoints [28319359-28319359]: Waiting: prepare-snapshot
2020-02-20T22:01:58.463 GCGB2 [Ledger INFO] Closed ledger: [seq=28319360, hash=4a23ad]
2020-02-20T22:02:08.136 GCGB2 [Herder INFO] Quorum information for 28319359 : {"agree":23,"delayed":0,"disagree":0,"fail_at":6,"hash":"341a0a","ledger":28319359,"missing":0,"phase":"EXTERNA
LIZE","validated":true}

2020-02-20T22:02:08.147 GCGB2 [Ledger INFO] Got consensus: [seq=28319361, prev=4a23ad, txs=117, ops=234, sv: [ txH: 232f01, ct: 1582236123, upgrades: [ ] ]]
2020-02-20T22:02:08.167 GCGB2 [Tx INFO] applying ledger 28319361 (txs:117, ops:234)
```
(nothing notable here other than a publish is happening)
SCP messages from that machine:
```
[
               1582236122,
               "{ENV@COINQVESTHongKong |  i: 28319361 | NOMINATE | D: 341a0a | X: {'[ SIGNED@COINQVESTHongKong txH: dcfa78, ct: 1582236122, upgrades: [ ] ]'} | Y: {} }",
               true
            ],
...
[
               1582236128,
               "{ENV@sdf_watcher1 |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               true
            ],
          [
               1582236128,
               "{ENV@keybase1 |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               true
            ],

            [
               1582236128,
               "{ENV@LOBSTR1Europe |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               true
            ],
            [
               1582236128,
               "{ENV@keybase2 |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               true
            ],
            [
               1582236128,
               "{ENV@COINQVESTGermany |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               true
            ],
            [
               1582236128,
               "{ENV@COINQVESTFinland |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               true
            ],
            [
               1582236128,
               "{ENV@COINQVESTHongKong |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               true
            ],
            [
               1582236128,
               "{ENV@LOBSTR5Australia |  i: 28319361 | EXTERNALIZE | c: (1,[ txH: 232f01, ct: 1582236123, upgrades: [ ] ]) | h.n: 1 | (lastD): 341a0a }",
               true
            ],
```

Node at the time was directly connected to a few other validators:
"COINQVESTGermany"
"wirexUS"
"SatoshiPayDEFrankfurt"
"keybase_io"
"wirexUK"
"sdf_watcher2"
"sdf_watcher3"

What is interesting here:
* clocks are the same on machines, so we can compare timestamps
* test machine received EXTERNALIZE messages at 1582236125
* sdf node received the same messages at 1582236128, even for validators that it's directly connected with

**Comment by MonsieurNicolas** [comment]
An interesting pattern I am seeing is that `messageSender` doesn't seem to batch things much:
as the app "cranks" fast enough, it's issuing single buffer writes most of the time.

Here's an example of what the logs (with https://github.com/stellar/stellar-core/pull/2442) give me (51 single writes in a second for that peer):
```
2020-02-22T01:51:05.081 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.086 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.093 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.099 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.127 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.186 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.207 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.208 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.209 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:268 n:1/1
2020-02-22T01:51:05.211 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:260 n:1/1
2020-02-22T01:51:05.216 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.220 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.224 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.232 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.237 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.241 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.284 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.311 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
2020-02-22T01:51:05.345 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:268 n:1/1
2020-02-22T01:51:05.350 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:260 n:1/1
2020-02-22T01:51:05.381 GCJMW [Overlay DEBUG] messageSender 34.80.95.189:11625 - b:444 n:1/1
```

If this is the case, we're missing out on aggregation quite a bit on both ends...

I also run network captures to see how the traffic looks like *outgoing* (as incoming we're connected to old peers), and indeed most of the traffic we have packets less than 100 bytes.

I compared 12.3.0 to master (12.4.0rc1) and the traffic stats appear to be about the same.

What I am seeing (using the "Throughput" analyzer in Wireshark) is that 12.4.0rc1 is a lot better than 12.3.0 at *receiving* data but about the same at *sending* data.

12.4.0rc1 receiving is
![image](https://user-images.githubusercontent.com/4213763/75086891-31a3f600-54ee-11ea-92e2-fcf1b037d6a2.png)

12.4.0rc1 sending is
![image](https://user-images.githubusercontent.com/4213763/75086886-23ee7080-54ee-11ea-992b-28ddb699544f.png)

12.3.0 receiving is
![image](https://user-images.githubusercontent.com/4213763/75086900-43859900-54ee-11ea-88a9-97f3e5ff318c.png)

12.3.0 sending is
![image](https://user-images.githubusercontent.com/4213763/75086904-57c99600-54ee-11ea-8997-45eb9a24a1ce.png)

So what I suspect is happening with 12.4.0 is that the overlay debt between what is being received vs sent is bigger, which contributes to the overall performance being worst.

**Comment by graydon** [comment]
Yeah, it's possible -- I saw some of this sort of pattern too when looking up close, though I did also see good aggregate numbers in broader simulations. One thing to keep in mind is that the scatter/gather change got rid of buffered writing altogether (we used to use the asio buffered _writer_ as well as reader; the write side was removed to use scatter/gather) so it might actually be writing _worse_ overall, if we're not batching enough into a scatter/gather set. I will do some experiments on this tomorrow.

**Comment by MonsieurNicolas** [comment]
I am going to leave it for now.

I updated the PR with a few more changes (based on your feedback), and added sqlite changes to avoid constant stalls (instead it's one stall slightly longer every once in a while).

I tried a few other things with surprising effect (branch https://github.com/MonsieurNicolas/stellar-core/tree/overlayBatchConfigsBuffered ):
* I added a delay of 100ms to force grouping instead of relying on the asio lottery
* I re-added the write buffer + flush (and according to logs it's flushing a good amount of data at once)

But those changes didn't have much effect on the on-wire traffic: it's still mostly small writes. Maybe I have not been doing the captures properly?
I have been capturing with `tcpdump -nnS port 11625 -w /tmp/overlay-12.3.0.cap`

I also tried to re-enable Nagle for all tests I did but that didn't have any effect. I am pretty sure that the other side still having it disabled should not matter.

**Comment by graydon** [comment]
I spent the day digging in ASIO's buffered stream class and measuring syscalls, pursuing a simpler hypothesis -- that we're delaying writes too much by spending too long in the new synchronous read-dispatch loop -- and I at least got some half-baked evidence that it might help to limit that (PR https://github.com/stellar/stellar-core/pull/2446 caps the iteration count at 10, which appears to knock about 30% off the p99 of scp.timing.externalized in simulation).

I did some experiments locally with it capped at _1_ (i.e. always bouncing off ASIO as before, but still using the increased read buffer size so we don't go to the OS as often, and still doing vectored writes to handle higher load) and I would like to try this in simulation later on but I am out of time tonight.

**Comment by MonsieurNicolas** [comment]
cool, so basically setting 1ms max time and 10 items. 1ms seems super aggressive but tuning those numbers down at worst gets us to where we were before (with some perf gain from the increased buffers).
If we merge #2442 we can try a few variations without having to rebuild everytime (and we get better logging), I updated the PR

**Comment by graydon** [comment]
@MonsieurNicolas I approved the PR and will run a simulation test or two using it later; ultimately I think we don't really want these to be exposed configs (they're Yet More Knobs To Turn To Wrong Settings) but for the sake of experiment and gaining confidence / adjusting to the right values I agree it'll make for faster turnaround.

**Comment by MonsieurNicolas** [comment]
I am still observing this. With the new metric `scp.timing.externalize_lag`, I am seeing this on a test node waiting on 2 out of 3 of `[sdf1, sdf2, sdf3]` (what we see here is the P99, the P75 is at 0):

![image](https://user-images.githubusercontent.com/4213763/77021332-a3cbf700-6943-11ea-85a5-16c13ea71b53.png)

Zoomed in:
![image](https://user-images.githubusercontent.com/4213763/77020987-b691fc00-6942-11ea-8e07-796fb922c369.png)

Corresponding `scp.timing.externalize_delay`:

![image](https://user-images.githubusercontent.com/4213763/77021360-b9412100-6943-11ea-9d3b-ee0ff39649a4.png)


![image](https://user-images.githubusercontent.com/4213763/77021277-7bdc9380-6943-11ea-9175-c5079acdbc0f.png)


Basically, we're seeing delays of 2-3 seconds relatively often (about once every 5 minutes), with gaps of about 1 minute where there are no delays (ie, we receive at least 2 messages at once).

Note the spikes for "delayed" (up to 20 seconds!), implying that we're receiving the last message a lot later even though all 3 sdf nodes are externalizing pretty much at the same time.
